my_credit_project

# TABLE OF CONTENT
1. Exploring Dataset
2. Data Preprocessing
3. Answering Questions
        Is there a connection between having kids and repaying a loan on time?
        Is there a connection between marital status and repaying a loan on time?
        Is there a connection between income level and repaying a loan on time?
        How do different loan purposes affect on-time loan repayment?

The purpose of this report is to investigate whether a customer's marital status and number of children have an impact on their likelihood of defaulting on a loan. 

The findings from this analysis will be used by the bank's loan division to enhance their credit scoring system, which evaluates the creditworthiness of potential borrowers.

# Exploring Dataset

In [6]:
import pandas as pd
credit_project = pd.read_csv('credit_scoring_eng.csv')
credit_project


children  days_employed  dob_years            education  education_id  \
0             1   -8437.673028         42    bachelor's degree             0   
1             1   -4024.803754         36  secondary education             1   
2             0   -5623.422610         33  Secondary Education             1   
3             3   -4124.747207         32  secondary education             1   
4             0  340266.072047         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1   -4529.316663         43  secondary education             1   
21521         0  343937.404131         67  secondary education             1   
21522         1   -2113.346888         38  secondary education             1   
21523         3   -3112.481705         38  secondary education             1   
21524         2   -1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose  
0         40620.102    purchase of the house  
1         17932.802             car purchase  
2         23341.752    purchase of the house  
3         42820.568  supplementary education  
4         25378.572        to have a wedding  
...             ...                      ...  
21520     35966.698     housing transactions  
21521     24959.969        purchase of a car  
21522     14347.610                 property  
21523     39054.888        buying my own car  
21524     13127.587             to buy a car  

[21525 rows x 12 columns]

In [8]:
# try loading the data from my local computer 
try:
    credit_project = pd.read_csv('credit_scoring_eng.csv')
# if error exist use the remote path in Jupyterhub
except:
    credit_project = pd.read_csv('/datasets/credit_scoring_eng.csv')

In [10]:
num_rows, num_columns = credit_project.shape
print('Number of rows:', num_rows,
      '\nNumber of columns:', num_columns)

Number of rows: 21525 
Number of columns: 12


In [11]:
credit_project.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

In [12]:
credit_project.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Based on the info above, there are missing values in days_employed and total_income columns 
and these columns are of float data type. 

The data types that we are dealing with are floating numbers, integer, values, and strings.

In [18]:
credit_project.head()

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

Based on the table above, the education column have implicit duplicates. 
Additionally, the days employed column has negative values which should be an issue since days cannot be less than 0.

In [20]:
# calculating the number of missing values
credit_project.isna().sum()


children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

# Data Preprocessing

In [26]:
# looking in the filtered table at the the first column with missing data
credit_project.loc[credit_project['days_employed'].isna()]

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

The missing values appear to be symmetric, 
but we need to verify if the missing values are symmetric for both 'days_employed' and 'total_income' columns.

In [25]:
# investigating if the missing values in both columns are symmetric
credit_project.loc[(credit_project['days_employed'].isna()) & (credit_project['total_income'].isna())]

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  secondary education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    bachelor's degree             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      m        retiree     0   
26               married                 0      m  civil servant     0   
29             unmarried                 4      f        retiree     0   
41               married                 0      f  civil servant     0   
55     civil partnership                 1      f        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      m       business     0   
21495  civil partnership                 1      f       employee     0   
21497            married                 0      f       business     0   
21502            married                 0      f       employee     0   
21510            married                 0      f       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

Intermediate conclusion

Since the number of rows is the same (2174 rows), 
the missing values in the 'days_employed' column and the 'total_income' column are symmetric. 
If a row has a missing value in the 'days_employed' column, 
it will also have a missing value in the 'total_income' column.

To further analyze the data, additional investigation needs to be done to determine if any patterns exist related to the missing values.

The next steps would be:

1. Calculating the percentage of missing values compared to the whole dataset. 
    This will help determine if the missing values constitute a significant portion of the data.
2. Investigating the distribution of the missing values in the 'days_employed' and 'total_income' columns. 
    This analysis may reveal any underlying reasons for the missing values.

In [27]:
# percentage of missing values

print(f"The percentage of missing values compared to the whole dataset is: {(credit_project['days_employed'].isna().sum()/num_rows*100):.2f}%")

The percentage of missing values compared to the whole dataset is: 10.10%


In [30]:
# removing implicit data by converting all data to lowercase

credit_project = credit_project.applymap(lambda x: x.lower() if isinstance(x, str) else x)

# investigating clients who do not have data on identified characteristic and the column with the missing values

credit_project_nan = credit_project.loc[credit_project['days_employed'].isna()].reset_index(drop=True)

credit_project_nan

children  days_employed  dob_years            education  education_id  \
0            0            NaN         65  secondary education             1   
1            0            NaN         41  secondary education             1   
2            0            NaN         63  secondary education             1   
3            0            NaN         50  secondary education             1   
4            0            NaN         54  secondary education             1   
...        ...            ...        ...                  ...           ...   
2169         2            NaN         47  secondary education             1   
2170         1            NaN         50  secondary education             1   
2171         0            NaN         48    bachelor's degree             0   
2172         1            NaN         42  secondary education             1   
2173         2            NaN         28  secondary education             1   

          family_status  family_status_id gender    income_type  debt  \
0     civil partnership                 1      m        retiree     0   
1               married                 0      m  civil servant     0   
2             unmarried                 4      f        retiree     0   
3               married                 0      f  civil servant     0   
4     civil partnership                 1      f        retiree     1   
...                 ...               ...    ...            ...   ...   
2169            married                 0      m       business     0   
2170  civil partnership                 1      f       employee     0   
2171            married                 0      f       business     0   
2172            married                 0      f       employee     0   
2173            married                 0      f       employee     0   

      total_income                   purpose  
0              NaN         to have a wedding  
1              NaN                 education  
2              NaN    building a real estate  
3              NaN  second-hand car purchase  
4              NaN         to have a wedding  
...            ...                       ...  
2169           NaN         purchase of a car  
2170           NaN          wedding ceremony  
2171           NaN       building a property  
2172           NaN    building a real estate  
2173           NaN              car purchase  

[2174 rows x 12 columns]

In [32]:
# checking for any remaining implicit data before analyzing the data distribution
for column in credit_project.columns:
    if credit_project[column].dtype == 'object':
        unique_values = credit_project[column].sort_values().unique()
        print(f"Unique values for column '{column}':")
        print(unique_values)
        print()

Unique values for column 'education':
["bachelor's degree" 'graduate degree' 'primary education'
 'secondary education' 'some college']

Unique values for column 'family_status':
['civil partnership' 'divorced' 'married' 'unmarried' 'widow / widower']

Unique values for column 'gender':
['f' 'm' 'xna']

Unique values for column 'income_type':
['business' 'civil servant' 'employee' 'entrepreneur'
 'paternity / maternity leave' 'retiree' 'student' 'unemployed']

Unique values for column 'purpose':
['building a property' 'building a real estate'
 'buy commercial real estate' 'buy real estate'
 'buy residential real estate' 'buying a second-hand car'
 'buying my own car' 'buying property for renting out' 'car'
 'car purchase' 'cars' 'construction of own property' 'education'
 'getting an education' 'getting higher education' 'going to university'
 'having a wedding' 'housing' 'housing renovation' 'housing transactions'
 'profile education' 'property' 'purchase of a car'
 'purchase of my ow

In [35]:
# distribution for each column in filtered data
for column in credit_project_nan.columns:
    if credit_project_nan['days_employed'].isnull().any():
        print(f"Distribution of values for column '{column}':")
        print(credit_project_nan[column].value_counts().sort_index())
        print()
      

Distribution of values for column 'children':
-1        3
 0     1439
 1      475
 2      204
 3       36
 4        7
 5        1
 20       9
Name: children, dtype: int64

Distribution of values for column 'days_employed':
Series([], Name: days_employed, dtype: int64)

Distribution of values for column 'dob_years':
0     10
19     1
20     5
21    18
22    17
23    36
24    21
25    23
26    35
27    36
28    57
29    50
30    58
31    65
32    37
33    51
34    69
35    64
36    63
37    53
38    54
39    51
40    66
41    59
42    65
43    50
44    44
45    50
46    48
47    59
48    46
49    50
50    51
51    50
52    53
53    44
54    55
55    48
56    54
57    56
58    56
59    34
60    39
61    38
62    38
63    29
64    37
65    20
66    20
67    16
68     9
69     5
70     3
71     5
72     2
73     1
Name: dob_years, dtype: int64

Distribution of values for column 'education':
bachelor's degree       544
primary education        21
secondary education    1540
some college     

Based on the distribution information provided above, 
no specific pattern or reason has been identified for the missing values.

In [36]:
# checking the distribution in the whole dataset

for column in credit_project.columns:
        print(f"Distribution of values for column '{column}':")
        print(credit_project[column].value_counts().sort_index())
        print()

Distribution of values for column 'children':
-1        47
 0     14149
 1      4818
 2      2055
 3       330
 4        41
 5         9
 20       76
Name: children, dtype: int64

Distribution of values for column 'days_employed':
-18388.949901     1
-17615.563266     1
-16593.472817     1
-16264.699501     1
-16119.687737     1
                 ..
 401663.850046    1
 401674.466633    1
 401675.093434    1
 401715.811749    1
 401755.400475    1
Name: days_employed, Length: 19351, dtype: int64

Distribution of values for column 'dob_years':
0     101
19     14
20     51
21    111
22    183
23    254
24    264
25    357
26    408
27    493
28    503
29    545
30    540
31    560
32    510
33    581
34    603
35    617
36    555
37    537
38    598
39    573
40    609
41    607
42    597
43    513
44    547
45    497
46    475
47    480
48    538
49    508
50    514
51    448
52    484
53    459
54    479
55    443
56    487
57    460
58    461
59    444
60    377
61    355
62    352
63

The distribution in the original dataset is similar to the distribution of the filtered table. 
This suggests that the missing values are missing randomly and 
there is no significant bias or specific pattern associated with the missing values.

In [37]:
# checking the unique values in the `children` column
credit_project['children'].sort_values().unique()

array([-1,  0,  1,  2,  3,  4,  5, 20], dtype=int64)

There are 2 issues that need to be addressed:

1. The number of children should not be negative. It is possible that the negative values are typos and should actually be positive values.
2. The value of 20 is an outlier, which seems to be incorrect. It is more likely that it should be 2 instead.

To resolve these issues, the negative values in the number of children column can be corrected by considering them as positive values. Similarly, the outlier value of 20 can be replaced with 2.

In [39]:
# replacing the wrong values

credit_project['children'] = credit_project['children'].replace({-1:1, 20:2})

In [41]:
# rechecking the unique values in the `children` column

credit_project['children'].sort_values().unique()

array([0, 1, 2, 3, 4, 5], dtype=int64)

In [42]:
# finding problematic data in `days_employed`, if they exist, and calculate the percentage

credit_project['days_employed'].value_counts().sort_index()

-18388.949901     1
-17615.563266     1
-16593.472817     1
-16264.699501     1
-16119.687737     1
                 ..
 401663.850046    1
 401674.466633    1
 401675.093434    1
 401715.811749    1
 401755.400475    1
Name: days_employed, Length: 19351, dtype: int64

In [45]:
# calculating percentage of problematic data

len(credit_project.loc[credit_project['days_employed'] < 0]) / len(credit_project)

0.7389547038327526

The proportion of negative values in the 'days_employed' column is relatively high, approaching 74%. 
This indicates that there is an issue with the data. 
Since it is impossible for days of employment to be negative, 
the values in the column will be converted to absolute values to ensure they are positive.

In [49]:
# converting the values in `days_employed` column to absolute values

credit_project['days_employed'] = credit_project['days_employed'].abs()

# rechecking the values in the `days_employed` column

len(credit_project.loc[credit_project['days_employed'] < 0]) / len(credit_project)

0.0

In [51]:
# converting 'days_employed' to years and add as column

credit_project['years_employed'] = credit_project['days_employed'] / 365
credit_project

children  days_employed  dob_years            education  education_id  \
0             1    8437.673028         42    bachelor's degree             0   
1             1    4024.803754         36  secondary education             1   
2             0    5623.422610         33  secondary education             1   
3             3    4124.747207         32  secondary education             1   
4             0  340266.072047         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1    4529.316663         43  secondary education             1   
21521         0  343937.404131         67  secondary education             1   
21522         1    2113.346888         38  secondary education             1   
21523         3    3112.481705         38  secondary education             1   
21524         2    1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      f    employee     0   
1                married                 0      f    employee     0   
2                married                 0      m    employee     0   
3                married                 0      m    employee     0   
4      civil partnership                 1      f     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      f    business     0   
21521            married                 0      f     retiree     0   
21522  civil partnership                 1      m    employee     1   
21523            married                 0      m    employee     1   
21524            married                 0      f    employee     0   

       total_income                  purpose  years_employed  
0         40620.102    purchase of the house       23.116912  
1         17932.802             car purchase       11.026860  
2         23341.752    purchase of the house       15.406637  
3         42820.568  supplementary education       11.300677  
4         25378.572        to have a wedding      932.235814  
...             ...                      ...             ...  
21520     35966.698     housing transactions       12.409087  
21521     24959.969        purchase of a car      942.294258  
21522     14347.610                 property        5.789991  
21523     39054.888        buying my own car        8.527347  
21524     13127.587             to buy a car        5.437007  

[21525 rows x 13 columns]

In [52]:
# checking if there is a duration of employment in years greater than the age

len(credit_project.loc[credit_project['years_employed'] >= credit_project['dob_years']]) / len(credit_project)

0.1634843205574913

In [53]:
# listing out the errors in 'years_employed' column

credit_project.loc[credit_project['years_employed'] >= credit_project['dob_years']]

children  days_employed  dob_years            education  education_id  \
4             0  340266.072047         53  secondary education             1   
18            0  400281.136913         53  secondary education             1   
24            1  338551.952911         57  secondary education             1   
25            0  363548.489348         67  secondary education             1   
30            1  335581.668515         62  secondary education             1   
...         ...            ...        ...                  ...           ...   
21505         0  338904.866406         53  secondary education             1   
21508         0  386497.714078         62  secondary education             1   
21509         0  362161.054124         59    bachelor's degree             0   
21518         0  373995.710838         59  secondary education             1   
21521         0  343937.404131         67  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
4      civil partnership                 1      f     retiree     0   
18       widow / widower                 2      f     retiree     0   
24             unmarried                 4      f     retiree     0   
25               married                 0      m     retiree     0   
30               married                 0      f     retiree     0   
...                  ...               ...    ...         ...   ...   
21505  civil partnership                 1      m     retiree     0   
21508            married                 0      m     retiree     0   
21509            married                 0      m     retiree     0   
21518            married                 0      f     retiree     0   
21521            married                 0      f     retiree     0   

       total_income                                   purpose  years_employed  
4         25378.572                         to have a wedding      932.235814  
18         9091.804                  buying a second-hand car     1096.660649  
24        46487.558  transactions with commercial real estate      927.539597  
25         8818.041                           buy real estate      996.023258  
30        27432.971  transactions with commercial real estate      919.401832  
...             ...                                       ...             ...  
21505     12070.399                         to have a wedding      928.506483  
21508     11622.175                                  property     1058.897847  
21509     11684.650                  real estate transactions      992.222066  
21518     24618.344                         purchase of a car     1024.645783  
21521     24959.969                         purchase of a car      942.294258  

[3519 rows x 13 columns]

In [54]:
credit_project['years_employed'].describe()

count    19351.000000
mean       183.328024
std        380.906522
min          0.066141
25%          2.539751
50%          6.011563
75%         15.172281
max       1100.699727
Name: years_employed, dtype: float64

The maximum value is unreasonably high and does not make sense. 
Additionally, there are a considerable number of error rows, accounting for 16% of the total. 
Therefore, the decision has been made to address these errors by replacing the erroneous values with the median.

In [55]:
# calculating the median of 'years_employed' data

years_employed_median = (credit_project.loc[credit_project['years_employed'] < credit_project['dob_years'], 'years_employed']).median()

years_employed_median

4.467647593196858

In [56]:
# converting the median of 'years_employed' to days

days_employed_median = years_employed_median * 365

days_employed_median

1630.6913715168532

In [58]:
# replacing the wrong values in 'years_employed' column with years_employed_median

credit_project.loc[credit_project['years_employed'] >= credit_project['dob_years'], 'years_employed'] = years_employed_median

In [59]:
len(credit_project.loc[credit_project['years_employed'] >= credit_project['dob_years']]) / len(credit_project)

0.004227642276422764

In [60]:
credit_project.loc[credit_project['years_employed'] >= credit_project['dob_years']]

children  days_employed  dob_years            education  education_id  \
99            0  346541.618895          0  secondary education             1   
149           0    2664.273168          0  secondary education             1   
270           3    1872.663186          0  secondary education             1   
578           0  397856.565013          0  secondary education             1   
1040          0    1158.029561          0    bachelor's degree             0   
...         ...            ...        ...                  ...           ...   
19371         1    2356.250615          0    bachelor's degree             0   
20462         0  338734.868540          0  secondary education             1   
20577         0  331741.271455          0  secondary education             1   
21179         2     108.967042          0    bachelor's degree             0   
21313         0    1268.487728          0  secondary education             1   

      family_status  family_status_id gender income_type  debt  total_income  \
99          married                 0      f     retiree     0     11406.644   
149        divorced                 3      f    employee     0     11228.230   
270         married                 0      f    employee     0     16346.633   
578         married                 0      f     retiree     0     15619.310   
1040       divorced                 3      f    business     0     48639.062   
...             ...               ...    ...         ...   ...           ...   
19371       married                 0      f    business     0     12469.795   
20462       married                 0      f     retiree     0     41471.027   
20577     unmarried                 4      f     retiree     0     20766.202   
21179       married                 0      m    business     0     38512.321   
21313     unmarried                 4      m    employee     0     24387.070   

                            purpose  years_employed  
99                              car        4.467648  
149            housing transactions        4.467648  
270              housing renovation        4.467648  
578    construction of own property        4.467648  
1040                   to own a car        4.467648  
...                             ...             ...  
19371                           car        4.467648  
20462      purchase of my own house        4.467648  
20577                      property        4.467648  
21179        building a real estate        4.467648  
21313          housing transactions        4.467648  

[91 rows x 13 columns]

It appears that we have fixed the incorrect values, except in cases where the age is 0. To fully address the remaining incorrect values, we need to prioritize fixing the instances where the age is 0.

In [61]:
# checking the `dob_years` for suspicious values and count the percentage

credit_project['dob_years'].sort_values().unique()

array([ 0, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51,
       52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68,
       69, 70, 71, 72, 73, 74, 75], dtype=int64)

In [62]:
# calculating percentage of problematic data

len(credit_project.loc[credit_project['dob_years'] == 0]) / len(credit_project)

0.004692218350754936

In [63]:
# dropping the problematic data in the `dob_years` column

credit_project = credit_project[credit_project['dob_years'] > 0]

# checking the result to make sure it's fixed

credit_project['dob_years'].sort_values().unique()

array([19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35,
       36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52,
       53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69,
       70, 71, 72, 73, 74, 75], dtype=int64)

Since the proportion of problematic values is less than 0.5%, 
the decision has been made to drop these values instead of filling them.

In [64]:
# calculating if the problematic data in 'years_employed' column still exist

len(credit_project.loc[credit_project['years_employed'] >= credit_project['dob_years']]) / len(credit_project)

0.0

In [66]:
# replacing the wrong values in 'days_employed' column with days_employed_median

credit_project.loc[credit_project['years_employed'] == years_employed_median, 'days_employed'] = days_employed_median

In [67]:
credit_project.loc[(credit_project['days_employed'] / 365) >= credit_project['dob_years']]

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose, years_employed]
Index: []

In [68]:
# checking if the error still exists

len(credit_project.loc[(credit_project['days_employed'] / 365) >= credit_project['dob_years']]) / len(credit_project)

0.0

In [69]:
# checking the unique values in the `family_status` column
credit_project['family_status'].sort_values().unique()

array(['civil partnership', 'divorced', 'married', 'unmarried',
       'widow / widower'], dtype=object)

In [70]:
credit_project['family_status'].value_counts().sort_index()

civil partnership     4156
divorced              1185
married              12331
unmarried             2797
widow / widower        955
Name: family_status, dtype: int64

The column 'family_status' does not contain any issues, so no correction is necessary.

In [72]:
# checking the `gender` column

credit_project['gender'].value_counts().sort_index()

f      14164
m       7259
xna        1
Name: gender, dtype: int64

In [76]:
# replacing 'xna' with the mode which is 'f'

credit_project['gender'] = credit_project['gender'].replace({'xna':'f'})


C:\Users\eyah\AppData\Local\Temp\ipykernel_27828\1042821676.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  credit_project['gender'] = credit_project['gender'].replace({'xna':'f'})


In [77]:
# checking the result to make sure it's fixed

credit_project['gender'].value_counts().sort_index()

f    14165
m     7259
Name: gender, dtype: int64

In [78]:
# checking the `income_type` column

credit_project['income_type'].value_counts().sort_index()

business                        5065
civil servant                   1453
employee                       11064
entrepreneur                       2
paternity / maternity leave        1
retiree                         3836
student                            1
unemployed                         2
Name: income_type, dtype: int64

The column 'income_type' does not contain any issues, so no correction is necessary.

In [79]:
# checking duplicates

credit_project.duplicated().sum()

71

In [80]:
# dropping the duplicates

credit_project = credit_project.drop_duplicates().reset_index(drop=True)

# last check whether we have any duplicates

credit_project.duplicated().sum()

0

In [82]:
# checking the size of the dataset after the first manipulations with it

credit_project.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21353 entries, 0 to 21352
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21353 non-null  int64  
 1   days_employed     19260 non-null  float64
 2   dob_years         21353 non-null  int64  
 3   education         21353 non-null  object 
 4   education_id      21353 non-null  int64  
 5   family_status     21353 non-null  object 
 6   family_status_id  21353 non-null  int64  
 7   gender            21353 non-null  object 
 8   income_type       21353 non-null  object 
 9   debt              21353 non-null  int64  
 10  total_income      19260 non-null  float64
 11  purpose           21353 non-null  object 
 12  years_employed    19260 non-null  float64
dtypes: float64(3), int64(5), object(5)
memory usage: 2.1+ MB


In [83]:
original_rows = 21525
after_rows = 21352

percentage_of_changes = (original_rows - after_rows)/original_rows*100
print(f"The percentage of changes is: {(percentage_of_changes):.2f}%")

The percentage of changes is: 0.80%


After correcting the dataset by dropping or replacing problematic data, the resulting dataset is reduced by only 0.80%, which is a negligible amount and will not significantly affect our analysis.

In [84]:
# finding the dictionaries

education_dict = credit_project.set_index('education').to_dict()['education_id']
family_dict = credit_project.set_index('family_status').to_dict()['family_status_id']
print(education_dict, family_dict)

{"bachelor's degree": 0, 'secondary education': 1, 'some college': 2, 'primary education': 3, 'graduate degree': 4} {'married': 0, 'civil partnership': 1, 'widow / widower': 2, 'divorced': 3, 'unmarried': 4}


Dictionaries are created to map and store relationships between key-value pairs 
and they can be used to map values from one form to another. 
The columns identified as identifiers are only 'education' and 'family_status'. 
Therefore, the dictionary created will only include values from these columns.

In [87]:
# checking data with missing values

credit_project.isna().sum()

children               0
days_employed       2093
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2093
purpose                0
years_employed      2093
dtype: int64

The columns with missing values are still 'days_employed' and 'total_income'. To address the issue of missing values in these columns, the following steps will be taken:

1. Age categories for clients will be created based on their age.

2. A new column will be added to the dataset to indicate the age category of each client. This column will provide additional information about the clients' age and can be used in further analysis or filling missing values.

By creating age categories and adding an age category column, we can gain more insights into the relationship between age and the missing values, which can help inform the handling of missing data in a more targeted manner.

In [88]:
# writing a function that calculates the age category

def age_category(age):
    if age <= 29:
        return '19-29'
    elif 29 < age <= 39:
        return '30-39'
    elif 39 < age <= 49:
        return '40-49'
    elif 49 < age <= 59:
        return '50-59'
    elif 59 < age <= 69:
        return '60-69'
    else:
        return '70-79'

In [89]:
# testing if the function works

print(age_category(21))
print(age_category(49))
print(age_category(77))
print(age_category(63))
print(age_category(59))
print(age_category(35))

19-29
40-49
70-79
60-69
50-59
30-39


In [90]:
# creating new column based on function

credit_project['age_category'] = credit_project['dob_years'].apply(age_category)

In [92]:
# checking how values in the new column

credit_project.head()

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0    1630.691372         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      f    employee     0     40620.102   
1            married                 0      f    employee     0     17932.802   
2            married                 0      m    employee     0     23341.752   
3            married                 0      m    employee     0     42820.568   
4  civil partnership                 1      f     retiree     0     25378.572   

                   purpose  years_employed age_category  
0    purchase of the house       23.116912        40-49  
1             car purchase       11.026860        30-39  
2    purchase of the house       15.406637        30-39  
3  supplementary education       11.300677        30-39  
4        to have a wedding        4.467648        50-59

In [93]:
# creating a table without missing values and print a few of its rows to make sure it looks fine

credit_project_wmv = credit_project.dropna()

credit_project_wmv

children  days_employed  dob_years            education  education_id  \
0             1    8437.673028         42    bachelor's degree             0   
1             1    4024.803754         36  secondary education             1   
2             0    5623.422610         33  secondary education             1   
3             3    4124.747207         32  secondary education             1   
4             0    1630.691372         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21348         1    4529.316663         43  secondary education             1   
21349         0    1630.691372         67  secondary education             1   
21350         1    2113.346888         38  secondary education             1   
21351         3    3112.481705         38  secondary education             1   
21352         2    1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      f    employee     0   
1                married                 0      f    employee     0   
2                married                 0      m    employee     0   
3                married                 0      m    employee     0   
4      civil partnership                 1      f     retiree     0   
...                  ...               ...    ...         ...   ...   
21348  civil partnership                 1      f    business     0   
21349            married                 0      f     retiree     0   
21350  civil partnership                 1      m    employee     1   
21351            married                 0      m    employee     1   
21352            married                 0      f    employee     0   

       total_income                  purpose  years_employed age_category  
0         40620.102    purchase of the house       23.116912        40-49  
1         17932.802             car purchase       11.026860        30-39  
2         23341.752    purchase of the house       15.406637        30-39  
3         42820.568  supplementary education       11.300677        30-39  
4         25378.572        to have a wedding        4.467648        50-59  
...             ...                      ...             ...          ...  
21348     35966.698     housing transactions       12.409087        40-49  
21349     24959.969        purchase of a car        4.467648        60-69  
21350     14347.610                 property        5.789991        30-39  
21351     39054.888        buying my own car        8.527347        30-39  
21352     13127.587             to buy a car        5.437007        40-49  

[19260 rows x 14 columns]

In [94]:
# checking the distribution for `age_category`

credit_project_wmv['age_category'].value_counts().sort_index()

19-29    2884
30-39    5109
40-49    4834
50-59    4178
60-69    2095
70-79     160
Name: age_category, dtype: int64

In [95]:
# checking the mean values for income based on `age_category`

print('mean based on',credit_project_wmv.groupby('age_category')['total_income'].mean(),'\n')

# checking the median values for income based on `age_category`

print('median based on',credit_project_wmv.groupby('age_category')['total_income'].median())

mean based on age_category
19-29    25533.960641
30-39    28312.479963
40-49    28551.375635
50-59    25811.700327
60-69    23242.812818
70-79    20125.658331
Name: total_income, dtype: float64 

median based on age_category
19-29    22742.6535
30-39    24667.5280
40-49    24764.2290
50-59    22203.0745
60-69    19817.4400
70-79    18751.3240
Name: total_income, dtype: float64


In [96]:
# checking the mean values for income based on `education`

print('mean based on',credit_project_wmv.groupby('education')['total_income'].mean(),'\n')

# checking the median values for income based on `education`

print('median based on',credit_project_wmv.groupby('education')['total_income'].median())

mean based on education
bachelor's degree      33172.428387
graduate degree        27960.024667
primary education      21144.882211
secondary education    24600.353617
some college           29040.391842
Name: total_income, dtype: float64 

median based on education
bachelor's degree      28054.5310
graduate degree        25161.5835
primary education      18741.9760
secondary education    21839.4075
some college           25618.4640
Name: total_income, dtype: float64


In [97]:
# checking the mean values for income based on `income_type`

print('mean based on',credit_project_wmv.groupby('income_type')['total_income'].mean(),'\n')

# checking the median values for income based on `income_type`

print('median based on',credit_project_wmv.groupby('income_type')['total_income'].median())

mean based on income_type
business                       32397.357125
civil servant                  27361.316126
employee                       25824.679592
entrepreneur                   79866.103000
paternity / maternity leave     8612.661000
retiree                        21939.310393
student                        15712.260000
unemployed                     21014.360500
Name: total_income, dtype: float64 

median based on income_type
business                       27564.8930
civil servant                  24083.5065
employee                       22815.1035
entrepreneur                   79866.1030
paternity / maternity leave     8612.6610
retiree                        18969.1490
student                        15712.2600
unemployed                     21014.3605
Name: total_income, dtype: float64


In [99]:
credit_project_wmv['total_income'].describe()

count     19260.000000
mean      26794.435857
std       16493.972601
min        3306.762000
25%       16494.872500
50%       23201.873500
75%       32539.207750
max      362496.645000
Name: total_income, dtype: float64

The data exhibits a skewed distribution and exhibits outliers, as indicated by the large difference between the 3rd quartile and the maximum value. Therefore, using the median as a measure of central tendency can be more robust. The median represents the middle value and is less influenced by extreme values. Consequently, filling missing values with the median would be more suitable in this scenario.

However, factors influencing the 'total_income' are not limited to the 'age_category' column alone. Both 'education' and 'income_type' can also have an impact on the clients' monthly income. Therefore, additional categorization based on these columns should be considered before filling in the missing values.


In [101]:
# writing a function that categorizes according to `age_category`, `education` and `income_type`

def special_category(age_category, education, income_type):
    for value in age_category:
        for value in education:
            for value in income_type:
                group = age_category + ", " + education + ", " + income_type
    return group

In [102]:
# testing if the function works

print(special_category('19-29', 'some college', 'employee'))

19-29, some college, employee


In [110]:
# creating new column based on function

credit_project['special_category'] = credit_project.apply(lambda row: special_category(row['age_category'], row['education'], row['income_type']), axis=1)

In [112]:
# recreating a table without missing values and missing values
credit_project_wmv = credit_project.dropna()
credit_project_nan = credit_project.loc[credit_project['days_employed'].isna()].reset_index(drop=True)

In [113]:
# checking the unique values for

print(len(set(credit_project_nan['special_category'])))
print(len(set(credit_project_wmv['special_category'])))

69
92


In [114]:
# checking if there are any values in `data_nan` that do not exist in `data_wmv`

values_not_in_credit_project_wmv = credit_project_nan[~credit_project_nan['special_category'].isin(credit_project_wmv['special_category'])]
values_not_in_credit_project_wmv

children  days_employed  dob_years          education  education_id  \
132         1            NaN         70  primary education             3   
560         0            NaN         58  bachelor's degree             0   
764         0            NaN         64  primary education             3   

         family_status  family_status_id gender    income_type  debt  \
132  civil partnership                 1      f       employee     0   
560            married                 0      m   entrepreneur     0   
764  civil partnership                 1      f  civil servant     0   

     total_income                                   purpose  years_employed  \
132           NaN  transactions with commercial real estate             NaN   
560           NaN               buy residential real estate             NaN   
764           NaN                         to have a wedding             NaN   

    age_category                         special_category  
132        70-79       70-79, primary education, employee  
560        50-59   50-59, bachelor's degree, entrepreneur  
764        60-69  60-69, primary education, civil servant

Since there are values in data_nan that do not exist in credit_project_wmv, 
it is not possible to fill in the missing values using the special_category created 
based on the matching values between the two datasets.

Now, the special_category column will be removed and a new category will be created.

In [115]:
# removing the `special_category` column from the table

credit_project.drop('special_category', axis=1, inplace=True)

credit_project.head()

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   
2         0    5623.422610         33  secondary education             1   
3         3    4124.747207         32  secondary education             1   
4         0    1630.691372         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      f    employee     0     40620.102   
1            married                 0      f    employee     0     17932.802   
2            married                 0      m    employee     0     23341.752   
3            married                 0      m    employee     0     42820.568   
4  civil partnership                 1      f     retiree     0     25378.572   

                   purpose  years_employed age_category  
0    purchase of the house       23.116912        40-49  
1             car purchase       11.026860        30-39  
2    purchase of the house       15.406637        30-39  
3  supplementary education       11.300677        30-39  
4        to have a wedding        4.467648        50-59

In [116]:
# writing a new function that categorizes according to `age_category` & `education`

def new_category(age_category, education):
    for value in age_category:
        for value in education:
            group = age_category + ", " + education
    return group

In [117]:
# creating new column based on the `new_category` function

credit_project['new_category'] = credit_project.apply(lambda row: new_category(row['age_category'], row['education']), axis=1)

In [118]:
# recreating a table without missing values and missing values

credit_project_nan = credit_project.loc[credit_project['days_employed'].isna()].reset_index(drop=True)
credit_project_wmv = credit_project.dropna()

In [119]:
print(len(set(credit_project_nan['new_category'])))
print(len(set(credit_project_wmv['new_category'])))

23
28


In [120]:
# checking again if there are any values in `data_nan` that do not exist in `data_wmv`

values_not_in_credit_project_wmv_2 = credit_project_nan[~credit_project_nan['new_category'].isin(credit_project_wmv['new_category'])]
values_not_in_credit_project_wmv_2

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose, years_employed, age_category, new_category]
Index: []

It appears that all the values for new_category in the data_nan dataset exist in the data_wmv dataset. 
Therefore, it is now possible to fill in the missing values using the 'new_category' column created.

In [121]:
# checking the median values for income based on `new_category`

print('median based on',credit_project_wmv.groupby('new_category')['total_income'].median())

median based on new_category
19-29, bachelor's degree      25956.1640
19-29, primary education      25488.9160
19-29, secondary education    21114.7620
19-29, some college           22687.1980
30-39, bachelor's degree      28794.9310
30-39, graduate degree        18187.3015
30-39, primary education      19542.3265
30-39, secondary education    22912.9930
30-39, some college           28463.4390
40-49, bachelor's degree      30282.3330
40-49, graduate degree        31771.3210
40-49, primary education      21511.5635
40-49, secondary education    22973.2580
40-49, some college           29323.6730
50-59, bachelor's degree      28152.1765
50-59, graduate degree        42945.7940
50-59, primary education      16922.6250
50-59, secondary education    21245.4820
50-59, some college           21132.3090
60-69, bachelor's degree      25222.3445
60-69, graduate degree        28334.2150
60-69, primary education      17657.4995
60-69, secondary education    18873.7640
60-69, some college         

In [123]:
credit_project_wmv.columns

Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose', 'years_employed', 'age_category',
       'new_category'],
      dtype='object')

In [124]:
#  writing a function that will be used for filling in missing values

def fill_missing_total_income(data, data_wmv):
    median_by_category = credit_project_wmv.groupby('new_category')['total_income'].median()
    credit_project['total_income'] = credit_project.apply(lambda row: median_by_category[row['new_category']] if pd.isnull(row['total_income']) else row['total_income'], axis=1)
    return credit_project
fill_missing_total_income(credit_project, credit_project_wmv)

children  days_employed  dob_years            education  education_id  \
0             1    8437.673028         42    bachelor's degree             0   
1             1    4024.803754         36  secondary education             1   
2             0    5623.422610         33  secondary education             1   
3             3    4124.747207         32  secondary education             1   
4             0    1630.691372         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21348         1    4529.316663         43  secondary education             1   
21349         0    1630.691372         67  secondary education             1   
21350         1    2113.346888         38  secondary education             1   
21351         3    3112.481705         38  secondary education             1   
21352         2    1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      f    employee     0   
1                married                 0      f    employee     0   
2                married                 0      m    employee     0   
3                married                 0      m    employee     0   
4      civil partnership                 1      f     retiree     0   
...                  ...               ...    ...         ...   ...   
21348  civil partnership                 1      f    business     0   
21349            married                 0      f     retiree     0   
21350  civil partnership                 1      m    employee     1   
21351            married                 0      m    employee     1   
21352            married                 0      f    employee     0   

       total_income                  purpose  years_employed age_category  \
0         40620.102    purchase of the house       23.116912        40-49   
1         17932.802             car purchase       11.026860        30-39   
2         23341.752    purchase of the house       15.406637        30-39   
3         42820.568  supplementary education       11.300677        30-39   
4         25378.572        to have a wedding        4.467648        50-59   
...             ...                      ...             ...          ...   
21348     35966.698     housing transactions       12.409087        40-49   
21349     24959.969        purchase of a car        4.467648        60-69   
21350     14347.610                 property        5.789991        30-39   
21351     39054.888        buying my own car        8.527347        30-39   
21352     13127.587             to buy a car        5.437007        40-49   

                     new_category  
0        40-49, bachelor's degree  
1      30-39, secondary education  
2      30-39, secondary education  
3      30-39, secondary education  
4      50-59, secondary education  
...                           ...  
21348  40-49, secondary education  
21349  60-69, secondary education  
21350  30-39, secondary education  
21351  30-39, secondary education  
21352  40-49, secondary education  

[21353 rows x 15 columns]

In [127]:
# checking missing values info
credit_project.isna().sum()

children               0
days_employed       2093
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
years_employed      2093
age_category           0
new_category           0
dtype: int64

In [128]:
# checking the mean values for work experience based on `age_category`

print('mean based on',credit_project_wmv.groupby('age_category')['days_employed'].mean(),'\n')

# checking the median values for work experience based on `age_category`

print('median based on',credit_project_wmv.groupby('age_category')['days_employed'].median())

mean based on age_category
19-29    1210.949536
30-39    2024.087792
40-49    2704.033064
50-59    2675.966154
60-69    2127.205365
70-79    1955.206065
Name: days_employed, dtype: float64 

median based on age_category
19-29     999.028882
30-39    1601.784231
40-49    1943.965054
50-59    1630.691372
60-69    1630.691372
70-79    1630.691372
Name: days_employed, dtype: float64


Generally, a person's work experience tends to increase with age. 
Older individuals, who have higher values in the 'dob_years' column, may have accumulated more work experience over the years. Hence, there is a positive correlation between 'days_employed' and 'dob_years', 
indicating that as age increases, the number of days employed also increases.

In [132]:
credit_project_wmv['days_employed'].describe()


count    19260.000000
mean      2225.039129
std       2107.888871
min         24.141633
25%        926.851509
50%       1630.691372
75%       2748.301767
max      18388.949901
Name: days_employed, dtype: float64

The data shows a skewed distribution with outliers, as evidenced by the significant difference between the 3rd quartile and the maximum value. 
In such cases, utilizing the median as a measure of central tendency can be more reliable. 
Unlike the mean, the median is not heavily influenced by extreme values, making it a more robust choice. 
Consequently, filling in missing values with the median would be a more appropriate approach in this particular scenario.

In [133]:
#  writing a function that will be used for filling in missing values

def fill_missing_days_employed(credit_project, credit_project_wmv):
    median_by_category_2 = credit_project_wmv.groupby('age_category')['days_employed'].median()
    credit_project['days_employed'] = credit_project.apply(lambda row: median_by_category_2[row['age_category']] if pd.isnull(row['days_employed']) else row['days_employed'], axis=1)
    return credit_project

fill_missing_days_employed(credit_project, credit_project_wmv)

children  days_employed  dob_years            education  education_id  \
0             1    8437.673028         42    bachelor's degree             0   
1             1    4024.803754         36  secondary education             1   
2             0    5623.422610         33  secondary education             1   
3             3    4124.747207         32  secondary education             1   
4             0    1630.691372         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21348         1    4529.316663         43  secondary education             1   
21349         0    1630.691372         67  secondary education             1   
21350         1    2113.346888         38  secondary education             1   
21351         3    3112.481705         38  secondary education             1   
21352         2    1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      f    employee     0   
1                married                 0      f    employee     0   
2                married                 0      m    employee     0   
3                married                 0      m    employee     0   
4      civil partnership                 1      f     retiree     0   
...                  ...               ...    ...         ...   ...   
21348  civil partnership                 1      f    business     0   
21349            married                 0      f     retiree     0   
21350  civil partnership                 1      m    employee     1   
21351            married                 0      m    employee     1   
21352            married                 0      f    employee     0   

       total_income                  purpose  years_employed age_category  \
0         40620.102    purchase of the house       23.116912        40-49   
1         17932.802             car purchase       11.026860        30-39   
2         23341.752    purchase of the house       15.406637        30-39   
3         42820.568  supplementary education       11.300677        30-39   
4         25378.572        to have a wedding        4.467648        50-59   
...             ...                      ...             ...          ...   
21348     35966.698     housing transactions       12.409087        40-49   
21349     24959.969        purchase of a car        4.467648        60-69   
21350     14347.610                 property        5.789991        30-39   
21351     39054.888        buying my own car        8.527347        30-39   
21352     13127.587             to buy a car        5.437007        40-49   

                     new_category  
0        40-49, bachelor's degree  
1      30-39, secondary education  
2      30-39, secondary education  
3      30-39, secondary education  
4      50-59, secondary education  
...                           ...  
21348  40-49, secondary education  
21349  60-69, secondary education  
21350  30-39, secondary education  
21351  30-39, secondary education  
21352  40-49, secondary education  

[21353 rows x 15 columns]

In [134]:
credit_project['days_employed'].isna().sum()

0

In [136]:
credit_project.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21353 entries, 0 to 21352
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21353 non-null  int64  
 1   days_employed     21353 non-null  float64
 2   dob_years         21353 non-null  int64  
 3   education         21353 non-null  object 
 4   education_id      21353 non-null  int64  
 5   family_status     21353 non-null  object 
 6   family_status_id  21353 non-null  int64  
 7   gender            21353 non-null  object 
 8   income_type       21353 non-null  object 
 9   debt              21353 non-null  int64  
 10  total_income      21353 non-null  float64
 11  purpose           21353 non-null  object 
 12  years_employed    19260 non-null  float64
 13  age_category      21353 non-null  object 
 14  new_category      21353 non-null  object 
dtypes: float64(3), int64(5), object(7)
memory usage: 2.4+ MB


In [137]:
# checking the unique values in the `purpose` column

credit_project['purpose'].sort_values().unique()

array(['building a property', 'building a real estate',
       'buy commercial real estate', 'buy real estate',
       'buy residential real estate', 'buying a second-hand car',
       'buying my own car', 'buying property for renting out', 'car',
       'car purchase', 'cars', 'construction of own property',
       'education', 'getting an education', 'getting higher education',
       'going to university', 'having a wedding', 'housing',
       'housing renovation', 'housing transactions', 'profile education',
       'property', 'purchase of a car', 'purchase of my own house',
       'purchase of the house', 'purchase of the house for my family',
       'real estate transactions', 'second-hand car purchase',
       'supplementary education', 'to become educated', 'to buy a car',
       'to get a supplementary education', 'to have a wedding',
       'to own a car', 'transactions with commercial real estate',
       'transactions with my real estate', 'university education',
       'we

In [139]:
# categorizing the 'purpose' column

def purpose_category(purpose):
    if 'property' in purpose:
        return 'property'
    elif 'real estate' in purpose:
        return 'real estate'
    elif 'car' in purpose:
        return 'car'
    elif 'educat' in purpose:
        return 'education'
    elif 'university' in purpose:
        return 'education'
    elif 'hous' in purpose:
        return 'house'
    elif 'wedding' in purpose:
        return 'wedding'
    else:
        return 'others'
    
credit_project['purpose_category'] = credit_project['purpose'].apply(purpose_category)

In [140]:
# checking if there are purposes categorized as others

credit_project.loc[credit_project['purpose_category'] == 'others']

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose, years_employed, age_category, new_category, purpose_category]
Index: []

In [141]:
credit_project

children  days_employed  dob_years            education  education_id  \
0             1    8437.673028         42    bachelor's degree             0   
1             1    4024.803754         36  secondary education             1   
2             0    5623.422610         33  secondary education             1   
3             3    4124.747207         32  secondary education             1   
4             0    1630.691372         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21348         1    4529.316663         43  secondary education             1   
21349         0    1630.691372         67  secondary education             1   
21350         1    2113.346888         38  secondary education             1   
21351         3    3112.481705         38  secondary education             1   
21352         2    1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      f    employee     0   
1                married                 0      f    employee     0   
2                married                 0      m    employee     0   
3                married                 0      m    employee     0   
4      civil partnership                 1      f     retiree     0   
...                  ...               ...    ...         ...   ...   
21348  civil partnership                 1      f    business     0   
21349            married                 0      f     retiree     0   
21350  civil partnership                 1      m    employee     1   
21351            married                 0      m    employee     1   
21352            married                 0      f    employee     0   

       total_income                  purpose  years_employed age_category  \
0         40620.102    purchase of the house       23.116912        40-49   
1         17932.802             car purchase       11.026860        30-39   
2         23341.752    purchase of the house       15.406637        30-39   
3         42820.568  supplementary education       11.300677        30-39   
4         25378.572        to have a wedding        4.467648        50-59   
...             ...                      ...             ...          ...   
21348     35966.698     housing transactions       12.409087        40-49   
21349     24959.969        purchase of a car        4.467648        60-69   
21350     14347.610                 property        5.789991        30-39   
21351     39054.888        buying my own car        8.527347        30-39   
21352     13127.587             to buy a car        5.437007        40-49   

                     new_category purpose_category  
0        40-49, bachelor's degree            house  
1      30-39, secondary education              car  
2      30-39, secondary education            house  
3      30-39, secondary education        education  
4      50-59, secondary education          wedding  
...                           ...              ...  
21348  40-49, secondary education            house  
21349  60-69, secondary education              car  
21350  30-39, secondary education         property  
21351  30-39, secondary education              car  
21352  40-49, secondary education              car  

[21353 rows x 16 columns]

# Answering Questions

The question is whether a customer's marital status and number of children have an impact on loan default. 

To firstly address and investigate the impact of marital status, 
we need to examine the columns family_status and debt.

Is there a connection between marital status and repaying a loan on time?

In [ ]:
# calculating the correlation between family_status and debt

credit_project[['family_status_id', 'debt']].corr()

In [143]:
# calculating the default rate on the type of marital status

credit_project.groupby('family_status_id').agg({'debt':['mean', 'median']})

debt       
                      mean median
family_status_id                 
0                 0.075427    0.0
1                 0.093462    0.0
2                 0.064990    0.0
3                 0.071730    0.0
4                 0.097709    0.0

The output shows the default rate (mean and median) for each 'family_status_id'. It appears that all the default rates are relatively low, and the median default rate for each marital status is 0, indicating that most individuals in each marital status category pay back their loans on time. 

Since the correlation value is less than 1, it supports the observation that there is no significant correlation between marital status and the likelihood of paying back a loan on time.

Now, let's address whether a customer's number of children have an impact on loan default.

Is there a connection between having kids and repaying a loan on time?

In [144]:
# calculating the correlation between number of children and debt

credit_project[['children', 'debt']].corr()

children      debt
children  1.000000  0.025028
debt      0.025028  1.000000

It appears that all the default rates based on number of children are also relatively low, and the median default rate for each number of children is 0, indicating that most individuals in each marital status category pay back their loans on time. Since the correlation value is approaching 0, it supports the observation that there is no significant correlation between customer's number of children and the likelihood of paying back a loan on time.

Next, let's analyze if there is a connection between income level and repaying a loan on time.

Is there a connection between income level and repaying a loan on time?

In [145]:
# calculating the correlation between total income and debt

credit_project[['total_income', 'debt']].corr()

total_income      debt
total_income      1.000000 -0.012984
debt             -0.012984  1.000000

Since the correlation is approaching 0, it indicates a weak or negligible correlation between income level and debt.

How do different loan purposes affect on-time loan repayment?

In [154]:
# perform one-hot encoding on the 'purpose_category' column
encoded_data = pd.get_dummies(credit_project, columns=['purpose_category'])

# calculating the correlation between the encoded columns and 'debt'
encoded_data.corr()['debt']

C:\Users\eyah\AppData\Local\Temp\ipykernel_27828\4236622524.py:5: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  encoded_data.corr()['debt']


children                        0.025028
days_employed                  -0.056416
dob_years                      -0.071553
education_id                    0.052751
family_status_id                0.020686
debt                            1.000000
total_income                   -0.012984
years_employed                 -0.059074
purpose_category_car            0.022402
purpose_category_education      0.020126
purpose_category_house         -0.023954
purpose_category_property      -0.008667
purpose_category_real estate   -0.010523
purpose_category_wedding       -0.001921
Name: debt, dtype: float64

The correlation coefficients for loans taken for car and education purposes exhibit very weak positive correlations. This indicates that acquiring a loan for a car or education has a slight positive influence on timely loan repayment, albeit the correlation is relatively weak.

On the other hand, loans taken for house and real estate purposes display very weak negative correlations. This implies that taking a loan for a house or real estate has an extremely minimal adverse effect on on-time loan repayment, despite the correlation being very weak.

Similarly, loans taken for property and wedding purposes exhibit extremely weak negative correlations. This signifies that obtaining a loan for property or a wedding has an exceedingly minimal negative impact on timely loan repayment, even though the correlation is very weak.

Therefore, the provided dataset suggests that there is no discernible connection between marital status, number of children, income level, and loan purposes with debt repayment.

The end.